Use more conservative circle detection

niklasmohrin opened the pull request:
Fixes #1503

This implements the exact logic for circle detection from `cat` for `bat`. To test this out yourself, you will have to get the newest version from the [pull request](https://github.com/niklasmohrin/clircle/pull/6) on the `clircle` repository and replace the `clircle` dependency in the `Cargo.toml` with the path to your local copy.

I will still have to fix the tests and see what to do with windows. Now that the posix approach (only error if program will provably fail later on) is taken here, we will have to investigate the situation on windows more closely.

I might still change things, there are breaking changes in `clircle` already, so there's no reason to hold back on changes over there.

Let me know, what you think and if you know anything about this on windows!

niklasmohrin commented:
About Windows: Now that circle detection works in the posix "only error out early if you know that there will be an error later" way, we should revisit what is done on Windows too. If I am honest, it does not look good for Windows users here.

- in `cmd.exe`, this problem does not exist like it does on Unix. The command `type a b > b` (or `>a`) loses the content of the output file, but the program does not go into an infinite loop (the current `bat` throws an error here)
- In Powershell, the cmdlet `Get-Content` does go into the infiinte loop trashing the disk, but it is impossible to detect this, because there is no such thing as redirecting output to a file. `>`  is syntax for piping into the `Out-File` cmdlet and the original program cannot find out, where the output is written to in the other process (naturally, `bat` does not detect this then).

In addition to that, now that `bat` behaves exactly as `cat`, if you run `bat a b > a`, no warning is shown, but the contents of `a` are lost.

@sharkdp What do you think about the idea of showing a warning, if `bat` detects a circle, but continues running because it can?

sharkdp commented:
Thank you for this work. I'd definitely like to include this `cat` compatibility change.

Concerning Windows: To be honest, I'm happy for any kind of support we can offer, but I'd even include this without any kind of detection on Windows. If we choose to do something, we should always try to be conservative and only show the error if we can be really sure.

niklasmohrin commented:
Commits "fix tests", tests are failing. How ironic.

@sharkdp I think I have it all set up. Feel free to also take a look at the PR in the other repository, as I have changed a lot of stuff. I also changed the default implementation of the `surely_conflicts_with` method to just return `false` to not produce any false positives on windows.

Does anybody know of the top of their head why access to the file that was just created is denied in the subprocess in these windows tests?

sharkdp commented on src/controller.rs:
```diff
@@ -67,13 +67,18 @@ impl<'b> Controller<'b> {
-                .flat_map(clircle::Identifier::try_from)
-                .collect();
-            clircle::stdout_among_inputs(&identifiers)
+        let stdout_identifier = if !attached_to_pager {
+            clircle::Identifier::try_from(clircle::Stdio::Stdout).ok()
+        } else {
+            None
```
Maybe reverse the two values and use the positive `if attached_to_pager { … }`

niklasmohrin replied:
Agreed.

Rust 1.50 would have had our back here with `(!attached_to_pager).then(clircle::Identifier::stdout)` (I will probably add the `Identifier::stdout` anyways) :laughing:

sharkdp replied:
Oh, that's a nice helper function :+1:

sharkdp commented on src/controller.rs:
```diff
@@ -67,13 +67,18 @@ impl<'b> Controller<'b> {
+        } else {
+            None
         };
+        // let rw_cycle_detected = !attached_to_pager && {
+        //     let identifiers: Vec<_> = inputs
+        //         .iter()
+        //         .flat_map(clircle::Identifier::try_from)
+        //         .collect();
+        //     clircle::stdout_among_inputs_relaxed(&identifiers)
+        // };
```
can be removed?

sharkdp commented on src/input.rs:
```diff
@@ -157,25 +157,55 @@ impl<'a> Input<'a> {
+                    let input_identifier = Identifier::try_from(clircle::Stdio::Stdin)
+                        .map_err(|e| format!("Stdin: Error identifying file: {}", e))?;
+                    if stdout.surely_conflicts_with(&input_identifier) {
+                        return Err(format!("Stdin: input file is output file",).into());
```
Should be easy to resolve this clippy suggestion

sharkdp commented on tests/integration_tests.rs:
```diff
@@ -209,21 +211,50 @@ fn line_range_multiple() {
 
+fn setup_cycle_file(filename: &str, content: &[u8]) -> PathBuf {
+    let path = PathBuf::from(EXAMPLES_DIR).join(filename);
+    let _ = File::create(path.as_path())
```
Remove the `let _ = ` part here.

sharkdp commented on tests/integration_tests.rs:
```diff
@@ -209,21 +211,50 @@ fn line_range_multiple() {
 #[test]
 fn basic_io_cycle() {
-    let file_out = Stdio::from(File::open("tests/examples/cycle.txt").unwrap());
+    let filename = setup_cycle_file("cycle1.txt", b"I am not empty");
```
This file is being created, but it's also checked in to the repository. Is that intended?

We usually create temporary directories if we really want to modify / create files for tests.

niklasmohrin replied:
I was gonna ask you about replacing `tempdir` with `tempfile` but just saw that it is already switched :smile: Yep, that is cleaner

sharkdp replied:
Maybe that would resolve the permission problems on Windows as well?

niklasmohrin replied:
I read the documentation of `tempfile` and since a temporary file would only give me the file and not a path to pass to `bat`, I ended up using directories after all. However, this did not fix the problems on Windows. I have actually read over the code again and found that another test shouldn't fail on Windows, so I added that in and obviously, that one also fails with a permission error. I have to do some more reading.

I first thought that the problem occurs, because temporary files may never be written to disk and the subprocess wouldn't be able to access them, but a `TempDir` just wraps a path to a normal directory (https://github.com/Stebalien/tempfile/blob/master/src/dir.rs#L403). It wouldn't explain why the tests failed with the files in `tests/examples` either.

This might take some more time :man_shrugging:

niklasmohrin replied:
After some intense sessions in the stdlib source code and msdn, I decided to add some `dbg!` stuff in my windows vm and to my surprise, it seemed that not the tempfile is the problem here, but `test.txt` (spoiler: it something completely different):

![image](https://user-images.githubusercontent.com/47574893/109216841-598b0100-77b5-11eb-8ef6-e7c8080f5d02.png)

(the thing to see here is that the error is printed before the `index = 1` which is the index in the main loop in `controller.rs`, so the error happens before the `-` argument or the temp file are even opened :scream: )

So I checked the security tab for `test.txt` and compared it to `tabs.txt`, but they both had the exact same list, so that can't it. I guess the access denied has nothing todo with reading after all ...

---

I narrowed it down to somewhere in `print_file`, so the access denied happens there somewhere. I will see what happens under hood when the process stdout gets pointed to a file and what sharing options are given and what not

niklasmohrin replied:
I found another small mistake, but I does not seem to be the problem. We could also consider making these tests unix exclusive. After all, we are _not doing anything_ on windows anyways. What do you think?

sharkdp replied:
Thank you for looking into this.

Running these tests on unix only is what I suggested below, right? I'd be okay with that. I'd also be okay with completely disabling `clircle` functionality on Windows.

sharkdp commented on tests/integration_tests.rs:
```diff
@@ -209,21 +211,50 @@ fn line_range_multiple() {
+    if cfg!(unix) {
+        ret.failure();
+    } else {
+        ret.success();
```
Do we really want to *enforce success* on other platforms. Or do we rather want to only  enforce failure on unix? In the latter case, just feature-gate the whole test with `#[cfg(unix)]`.

niklasmohrin replied:
What this test really does is check for the unix behavior, but it could also act as a test that checks that false positives don't happen on other platforms (or teach us about a platform that we missed in the implementation). I would be fine with only running it on unix too, what do you think?

sharkdp replied:
I'd prefer running it on unix only

sharkdp commented on src/input.rs:
```diff
@@ -157,25 +157,55 @@ impl<'a> Input<'a> {
+                    let input_identifier = Identifier::try_from(clircle::Stdio::Stdin)
+                        .map_err(|e| format!("Stdin: Error identifying file: {}", e))?;
+                    if stdout.surely_conflicts_with(&input_identifier) {
+                        return Err(format!("Stdin: input file is output file",).into());
```
I know it was similar before, but maybe we could extend this error message a little bit. After all, that's everything that a user is ever going to see from this feature. When is this error triggered? If someone calls `bat < file > file`?

niklasmohrin replied:
Good point, gotta beat coreutils in some way :sweat_smile:

sharkdp replied:
:smile:

sharkdp commented on src/input.rs:
```diff
@@ -157,25 +157,55 @@ impl<'a> Input<'a> {
+                        })?;
+                        if stdout.surely_conflicts_with(&input_identifier) {
+                            return Err(format!(
+                                "{}: input file is output file",
```
Similar here: maybe extend this to something like:
"I/O circle detected. One of the input files is also an output file. Aborting to avoid infinite loop."

sharkdp commented:
Thank you for the updates! It would be great if you could create a clircle release for this and update the PR here to use that release. From my point of view, this looks ready to go into [bat v0.18](https://github.com/sharkdp/bat/issues/1499).

sharkdp commented:
The clippy suggestions on Windows could be resolved by moving the `use` statements into the test functions.

niklasmohrin commented:
@sharkdp Sounds good! There is one more open thing: Since `surely_conflicts_with` is always `false` on windows, we could change the line that sets `stdout_identifier` to always be `None` on windows to not do the syscalls. What do you think?

sharkdp commented:
Sounds good!

niklasmohrin commented:
@sharkdp I think it's all done now

sharkdp reviewed (approved):
Thank you so much!
